crates/cpuid-utils/src/lib.rs

```diff
@@ -23,55 +63,475 @@ pub mod bits;
+                // The caller didn't supply a subleaf, but the leaf has one, so
+                // the keys don't match.
+                (None, Subleaves::Present(_)) => {
+                    return None;
+                }
+                // The caller supplied a subleaf, but the leaf doesn't use
+                // subleaves, so the keys don't match.
+                (Some(_), Subleaves::Absent(_)) => {
+                    return None;
+                }
+                // The caller supplied a subleaf, and this leaf has subleaf
+                // data. If the requested subleaf is in the leaf's subleaf map,
+                // remove the corresponding subleaf entry. If this empties the
+                // subleaf map, also clean up the leaf entry.
+                (Some(subleaf), Subleaves::Present(sl_map)) => {
+                    let val = sl_map.remove(&subleaf)?;
+                    (val, sl_map.is_empty())
+                }
+            }
+        };
+
+        if remove_leaf {
+            entry.remove_entry();
+        }
+
+        Some(val)
+    }
+
+    /// Removes all data for the supplied `leaf`. If the leaf has subleaves, all
+    /// their entries are removed.
+    pub fn remove_leaf(&mut self, leaf: u32) {
+        self.0.remove(&leaf);
+    }
+
+    /// Passes each leaf number in the map to `f` and removes any leaf entries
+    /// for which `f` returns `false`. If a removed leaf has subleaves, all
+    /// their entries are removed.
+    pub fn retain_by_leaf<F>(&mut self, mut f: F)
+    where
+        F: FnMut(u32) -> bool,
+    {
+        self.0.retain(|leaf, _| f(*leaf));
+    }
+
+    /// Clears the entire map.
+    pub fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Returns `true` if the map has no entries.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the total number of leaf/subleaf/value tuples in the map.
+    pub fn len(&self) -> usize {
+        let mut len = 0;
+        for sl in self.0.values() {
+            match sl {
+                Subleaves::Absent(_) => len += 1,
+                Subleaves::Present(sl_map) => len += sl_map.len(),
+            }
+        }
+
+        len
+    }
+
+    /// Returns an iterator over the ([`CpuidIdent`], [`CpuidValues`]) pairs in
+    /// the map.
+    pub fn iter(&self) -> CpuidMapIterator {
+        CpuidMapIterator::new(self)
+    }
+}
+
+/// An iterator over a [`CpuidMap`]'s leaf/subleaf/value tuples.
+pub struct CpuidMapIterator<'a> {
+    leaf_iter: btree_map::Iter<'a, u32, Subleaves>,
+    subleaf_iter: Option<(u32, btree_map::Iter<'a, u32, CpuidValues>)>,
+}
+
+impl<'a> CpuidMapIterator<'a> {
+    fn new(map: &'a CpuidMap) -> Self {
+        Self { leaf_iter: map.0.iter(), subleaf_iter: None }
+    }
+}
+
+impl Iterator for CpuidMapIterator<'_> {
+    type Item = (CpuidIdent, CpuidValues);
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
I appreciate the comments here --- the implementation makes sense, but they make it a lot clearer.